### ISSUE TITLE
DataType up/down with MySQL

### ISSUE THREAD

**Issue by hendrul** [sequelize/sequelize#5585]
Up/Down mecanism for datatypes (#4186) is not included in the docs yet?, I'm having troubles with mysql when parsing a BLOB from the database, it triggers my parse method even for TEXT fields and is not passing a Buffer as value as I expected, instead, it is passing an specific Field object from the mysql  driver into which I can't find the buffer contained and I have no clue about the type of the field from this object.

```javascript
//base64/Buffer converter for BLOB fields
Sequelize.BLOB.parse = function(value, options) { //This method is triggered even for TEXT fields
    if (Buffer.isBuffer(value)) {
        return value.toString('base64');
    } else if( typeof value.buffer == 'function' ) {
        //On MySQL I get a "Field" object as value instead of a Buffer
        //and I can't discern if value is BLOB or TEXT
        return value.buffer().toString('base64'); // buffer() doesn't work
        //value.parser._buffer has the buffer
        //Finally if I return the encoded base64 string, the queries misbehaves
    }
};

Sequelize.BLOB.prototype.stringify = function(value, options) {
    if (!Buffer.isBuffer(value)) {
        if (Array.isArray(value)) {
            value = new Buffer(value);
        } else if(validator.isBase64(value)){
            value = new Buffer(value.toString(), 'base64');
        } else if(validator.isHexadecimal(value)) {
            value = new Buffer(value.toString(), 'hex');
        } else {
            value = new Buffer(value.toString());
        }
    }
    var hex = value.toString('hex');
    return "X'" + hex + "'";
};
```

**Comment by janmeier** [comment]
Internally text fields are stored as BLOBs in mysql - thats why the parse function is trigged, see http://dev.mysql.com/doc/refman/5.7/en/blob.html

All parsing depends on the type the value has in the DB - not the type declared in the model. This has both ups and downs - The downside being that some types cannot be distinguished, the upside being that the parsing also works for raw queries without a model.

The field object you get is the raw data we get from node-mysql, which should definitely have a `buffer()` function, see https://github.com/felixge/node-mysql#type-casting.

The idea of the rewamped system is to allows users to hook more directly in the parsing system, that is why you get the raw object with `buffer`, `string` and `geometry` functions, as opposed to a buffer.

**Comment by hendrul** [comment]
> The field object you get is the raw data we get from node-mysql, which should definitely have a buffer() function

Yes it have a buffer function but it doesn't works as I would expect, the only way I see to get the complete buffer is value.parser._buffer.

> The downside being that some types cannot be distinguished

This is a terrible downside, distinguishing blobs and working on them is I think the most common use case for this feature, I think any other alternative (if they exists) is better than this. How can I do to convert up/down a blob to a base64 string representation using this feature?.

> The idea of the rewamped system is to allows users to hook more directly in the parsing system

Doesn't this makes client code DB specific? which is one of the things sequelize is trying (where possible) to abstract away from user?, Why sequelize doesn't get the buffer for me? and have me expect always to receive a buffer. If am running tests on SQLite and using any other database for production I should conditionally read this object based on what DB is selected.

Another anomally I saw is that even without TEXT fields on the DB, only BLOBs and reading the buffer (through value.parser._buffer) then returning its base64 string rep. causes the querys to be inconsistent... I'll give you the test snippet when as soon I get access to that code again.

**Comment by janmeier** [comment]
> Yes it have a buffer function but it doesn't works as I would expect, the only way I see to get the complete buffer is value.parser._buffer. 

How so?  We use the `buffer` function [when parsing geometry](https://github.com/sequelize/sequelize/blob/6486f3c36b469d1afe0bda8aef013275901adf47/lib/dialects/mysql/data-types.js#L80) and that seems to work fine 

> This is a terrible downside, distinguishing blobs and working on them is I think the most common use case for this feature, I think any other alternative (if they exists) is better than this. How can I do to convert up/down a blob to a base64 string representation using this feature?.

I agree, this is a big limitation, and means that the datatype system cannot fulfill its complete potential as I had envisioned it beforehand. The alternative would be to use the datatype from the model. However, this would require the calling method to pass a map of the type for each column (and each nested column when using includes) - it's definitely not impossible, but it would require some work. 

On top of that, if the column is an aggregate, then we still don't know the type - if you concatenate two text fields, it will show up as a buffer, and there is no way we can figure that out.

The best would probably be to have both solutions - use the type from the model definition if present, and if not fall back to the type given by the DB. This would make it possible to create truly custom datatypes - for example a type which is stored as a string in the DB, but has some custom parsing / serialization logic. But someone has to actually take the time to sit down and write that code :)

> Doesn't this makes client code DB specific? which is one of the things sequelize is trying (where possible) to abstract away from user?, Why sequelize doesn't get the buffer for me? and have me expect always to receive a buffer. If am running tests on SQLite and using any other database for production I should conditionally read this object based on what DB is selected.

It definitely does! Generally we try to hide the differences away yes, but in this case we actually wanted to give the user more direct access to parsing and serialization without trying to be clever. Writing code to hide to differences between the different APIs would be a lot of work, and it would be hard to even come up with a consistent API - since the databases are just genuinely different.

TL;DR - when you are using the datatype system you are hooking into some pretty internal parts - and thus you'll need to do a bit more work :)
